### PR TITLE
Add support for secret length in bytes and additional encodings

### LINKED ISSUE #27: Configurable encoding / byte-oriented units for length of encoded data
**Is your feature request related to a problem? Please describe.**

I have an application I want to generate a SHARED_SECRET for (Pomerium). Pomerium specifically needs a 256-bit key for this, and it expects that the value of the secret it reads will be a [Base64-encoded string representing 32 bytes](https://www.pomerium.io/reference/#shared-secret).

With secret-generator as it is currently constituted, I can't generate this key: either I set a length of `"40"` and Pomerium fails to start while complaining that 30 bytes is too short, I set a length of `"44"` and Pomerium fails to start while complaining that 33 bytes is too long, or I set a length between the two and Pomerium fails to start while complaining that Base64 is malformed.

**Describe the solution you'd like**

The main shortcoming of the current system underlying this problem is that the `length` always refers to the length of the generated *string*, not the length of the byte sequence *encoded by* that string. Even in situations where the underlying binary data of the secret doesn't matter, it can helpful to calculate strength in terms of bits of randomness.

To meet these needs, I propose recognizing a `B` suffix for the (yet-to-be-documented) `secret-generator.v1.mittwald.de/length` annotation to specify the length of the random secret in bytes, similar to how other resources can be specified in Kubernetes with a distinguishing suffix to change their meaning (such as the `m` suffix on CPU requests/limits, which specifies the CPU quota in terms of [millicores](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu)).

That resolves my current use case: however, there are many other applications for Secrets, and many of them expect their random secret data to be encoded in another form than a Base64 string (which is, itself, encoded into Base64 again in the `data` field of the `Secret`). As such, I also propose a `secret-generator.v1.mittwald.de/encoding` annotation, which can generate secrets for all of these applications:

- Systems that take key material directly from a file could use an encoding type of `raw` (as they are still encoded as Base64 internally by Kubernetes).
- Other encodings, many of which are already included in Go's core `encoding` routines, could be specified to use different representations for the underlying random bits, while also serving as a shorthand for the character set, if looking to just generate a random string (by specifying a non-`B`-suffixed length):
  - `base10` (digits)
  - `hex` / `base16` (hexadecimal)
  - `base32` (homograph-avoidant lowercase alphanumeric set)
  - `base36` (full lower-case alphanumeric set)
  - `base62` (full alphanumeric set, a la Sprig's `randAlphaNum`)
  - `base64url` (URL-safe characters)
  - `ascii85` (full ASCII, a la Sprig's `randAscii`)

**Describe alternatives you've considered**

The most flexible form of `encoding` would be to let the user specify their own "alphabet" for converting binary values to arbitrary digit systems. However, this much fine-tuned control over a random secret, at this level of the cluster, would most likely be the wrong solution to the wrong problem: even the most sensitive of differences can generally be solved with one of the encodings, or a common variation thereof (ie. Douglas Crockford's version of Base32).

For lengths, I originally considered a separate `unit` annotation that could be either `bytes` or `digits`. However, not only would the former be a lot of annotation bloat for a pretty trivial change, it would also result in a false compatibility: if a document needing a 32-byte base64-encoded string were deployed to a cluster where an older version of the secret-generator were deployed, rather than failing to generate the secret (as it wouldn't recognize the length) and potentially leaving that up to a different component of the system (ie. one that tries to unseal secrets that couldn't be randomly generated), it would *misinterpret* the length value, and generate a secret of the wrong size (which may even result in the secret being unparseable).

I also considered an alternative `bytes` annotation that would be mutually exclusive to `length`. However, on top of this also being a certain amount of annotation bloat (with a confusing distinction between the two annotations), this has no clear way to extend the annotation values if the value of `autogenerate` is a comma-separated list, where each key may have a different encoding or length. With the proposed design, both `encoding` and `length` could potentially be extended out to comma-separated lists in the future as well, where the first `length` or `encoding` refers to the first key of `autogenerate`, the second describing the second key, etc. (If this extension were implemented, their behavior when the lists are not the same length would need to be defined: such consideration is out of scope for this proposal.)

### PR DESCRIPTION
**Pull request by YannikBramkamp**
Closes #27

### CODE REVIEW

File: pkg/controller/secret/secret_controller.go [hunk 1]
```diff
@@ -151,14 +160,30 @@ func (r *ReconcileSecret) Reconcile(request reconcile.Request) (reconcile.Result
...
+    }
+	return l, byteLen, nil
+}
+
+func secretEncodingFromAnnotation(fallback string, annotations map[string]string) (string, error) {
+	l := fallback
+	if val, ok := annotations[AnnotationSecretEncoding]; ok {
+		l = val
 	}
 	return l, nil
```
Comment (martin-helmich) [inline pkg/controller/secret/secret_controller.go:188]: Shorter:

```suggestion
	if val, ok := annotations[AnnotationSecretEncoding]; ok {
		return l, nil
	}
	return fallback, nil
```

File: pkg/controller/secret/secret_controller.go [hunk 2]
```diff
@@ -151,14 +160,30 @@ func (r *ReconcileSecret) Reconcile(request reconcile.Request) (reconcile.Result
...
 }
 
-func secretLengthFromAnnotation(fallback int, annotations map[string]string) (int, error) {
+func secretLengthFromAnnotation(fallback int, annotations map[string]string) (int, bool, error) {
 	l := fallback
+	byteLen := false
+
 	if val, ok := annotations[AnnotationSecretLength]; ok {
-		intVal, err := strconv.Atoi(val)
+	val = strings.ToLower(val)
```
Comment (martin-helmich) [inline pkg/controller/secret/secret_controller.go:168]: Please `go fmt` everything. I'm wondering why Codeclimate didn't pick up on this... 🤔

File: pkg/controller/secret/secret_string.go [hunk 3]
```diff
@@ -39,7 +48,7 @@ func (pg StringGenerator) generateData(instance *corev1.Secret) (reconcile.Resul
 		}
 	}
 
-	length, err := secretLengthFromAnnotation(secretLength(), instance.Annotations)
+	length, byteLen, err := secretLengthFromAnnotation(secretLength(), instance.Annotations)
```
Comment (martin-helmich) [inline pkg/controller/secret/secret_string.go:51]: `byteLen` is named a bit ambiguously: From the name `byteLen`, one might believe that this variables contains "the actual desired secret length in bytes" and not "`true` if another variable `length` contains the byte length, else `false`". Maybe name this `isByteLength`?

File: pkg/controller/secret/secret_string.go [hunk 4]
```diff
@@ -73,14 +79,73 @@ func (pg StringGenerator) generateData(instance *corev1.Secret) (reconcile.Resul
...
+		}
+		instance.Data[key] = []byte(value)
+
+		pg.log.Info("set field of instance to new randomly generated instance", "bytes", len(value), "field", key, "encoding",encoding)
+	}
+
+	return nil
+}
+
+func generateRandomString(length int, encoding string, lenBytes bool) (string, error) {
```
Comment (martin-helmich) [inline pkg/controller/secret/secret_string.go:113]: You could make this function also return a `[]byte` (it's converted to `[]byte` later, anyway) and then make this function also handle the raw string generation (which is now done by `generateRandomStringRaw`)

File: pkg/controller/secret/types.go [hunk 5]
```diff
@@ -14,6 +14,7 @@ const (
 	AnnotationSecretType            = "secret-generator.v1.mittwald.de/type"
 	AnnotationSecretLength          = "secret-generator.v1.mittwald.de/length"
 	AnnotationBasicAuthUsername     = "secret-generator.v1.mittwald.de/basic-auth-username"
+	AnnotationSecretEncoding        = "secret-generator.v1.mittwald.de/encoding"
```
Comment (martin-helmich) [inline pkg/controller/secret/types.go:17]: Please don't forget to also adjust the `README.md`

File: pkg/controller/secret/secret_controller.go [hunk 6]
```diff
@@ -151,14 +160,30 @@ func (r *ReconcileSecret) Reconcile(request reconcile.Request) (reconcile.Result
 	return reconcile.Result{}, nil
 }
 
-func secretLengthFromAnnotation(fallback int, annotations map[string]string) (int, error) {
+func secretLengthFromAnnotation(fallback int, annotations map[string]string) (int, bool, error) {
 	l := fallback
+	byteLen := false
```
Comment (martin-helmich) [inline pkg/controller/secret/secret_controller.go:165]: See my other comment on the `byteLen` naming.

**Review by martin-helmich — CHANGES_REQUESTED** [review]

**Review by martin-helmich — APPROVED** [review]
LGTM! 👍

File: README.md [hunk 7]
```diff
@@ -64,6 +64,15 @@ This operator is capable of generating secure random strings and ssh keypair sec
 The type of secret to be generated can be specified by the `secret-generator.v1.mittwald.de/type` annotation.
 This annotation can be added to any Kubernetes secret object in the operators `watchNamespace`.
 
+The encoding of the secret can be specified by the `secret-generator.v1.mittwald.de/encoding` annotation.
+Available encodings are `base64`, `base64url`, `base32`, `hex` and `raw`, with `raw` returning the unencoded byte sequence
+that was generated. `base64` will be used, if the annotation was not used.
+
+The length of the generated secret can be specified by the `secret-generator.v1.mittwald.de/length` annotation.
+By default, this length refers to the length of the generated string, and not the length of the byte sequence encoded by it. 
+The suffix `B` or `b` can be used to indicate that the provided value should refer to the encoded byte sequence instead.
```
Comment (stuartpb) [inline README.md:73]: This is the one quibble I was hoping I wouldn't have to bring up by submitting the PR myself: the lower-case suffix `b` should **not** be supported, as `b` is the suffix for **bits**, not bytes, and should technically divide the value by 8 - value of `64b` is equivalent to `8B`. (This is more commonly recognized in forms like "Mbps" for "megabits per second", where the number of mega*bytes* that will be downloaded in a second must be calculating by dividing by eight).

**Review by stuartpb — COMMENTED** [review]
Looks great! I have a qualm with one minor detail, though - see comment inline. I didn't include this in my original spec from #27 because I felt it was rather wordy enough as it was :sweat_smile:

**Pull request merged**